language/stdlib/modules/Libra.move

```diff
@@ -700,14 +694,85 @@ module Libra {
+        pragma aborts_if_is_partial = true;
+        // aborts_if !Roles::spec_has_register_new_currency_privilege(lr_account);
+        include RegisterCurrencyAbortsIf<CoinType>;
+        ensures spec_is_currency<CoinType>();
+    }
+
+    spec schema RegisterCurrencyAbortsIf<CoinType> {
+        lr_account: signer;
+        // TODO (dd): Incomprehensible "function does not abort under this condition"
+        // aborts_if !Roles::spec_has_register_new_currency_privilege(lr_account);
         aborts_if Signer::spec_address_of(lr_account) != CoreAddresses::SPEC_CURRENCY_INFO_ADDRESS();
         aborts_if !exists<CurrencyRegistrationCapability>(CoreAddresses::SPEC_LIBRA_ROOT_ADDRESS());
         aborts_if exists<CurrencyInfo<CoinType>>(Signer::spec_address_of(lr_account));
         aborts_if spec_is_currency<CoinType>();
         include RegisteredCurrencies::AddCurrencyCodeAbortsIf;
+    }
 
-        ensures spec_is_currency<CoinType>();
+    /// Registers a stable currency (SCS) coin -- i.e., a non-synthetic currency.
+    /// Registration is a two-step process because resources are published on two distinct
+    /// accounts: The CoinInfo is published on the Libra root account, and the mint and
+    /// burn capabilities are published on a treasury compliance account.
+    /// `register_SCS_currency_start` does the first step by publishing the CoinInfo,
+    /// then publishing `Offer`s to `tc_addr` (an address of a treasury compliance account)
+    /// for the mint and burn capabilities.
+    ///
+    /// The second step, `register_SCS_currency_finish` is called with a
+    /// treasury compliance signer, who redeems the offer and publishes the capabilities.
+    /// This leaves open the possibility that different currencies can have different
+    /// treasury compliance accounts.
+    public fun register_SCS_currency_start<CoinType>(
+        lr_account: &signer,
+        tc_addr: address,
+        to_lbr_exchange_rate: FixedPoint32,
+        scaling_factor: u64,
+        fractional_part: u64,
+        currency_code: vector<u8>,
+    ) acquires CurrencyRegistrationCapability {
+        let (mint_cap, burn_cap) =
+            register_currency<CoinType>(
+                lr_account,
+                to_lbr_exchange_rate,
+                false,   // is_synthetic
+                scaling_factor,
+                fractional_part,
+                currency_code,
+            );
+        // Put capabilities in an offer for treasury account
+        Offer::create<BurnCapability<CoinType>>(lr_account, burn_cap, tc_addr);
+        Offer::create<MintCapability<CoinType>>(lr_account, mint_cap, tc_addr);
+    }
+
+    spec fun register_SCS_currency_start {
+        // TODO (dd): I could not figure out what the problem was here.
+        pragma aborts_if_is_partial = true;
+        include RegisterCurrencyAbortsIf<CoinType>;
+    }
+
+    public fun register_SCS_currency_finish<CoinType>(
```

Comment:
What goes wrong if we combine `start` and `finish` (given that they are always called together)? I am on board with getting rid of the `publish` functions for the mint and burn capabilities, but splitting this into two public procs + adding the indirection of `Offer` seems unfortunate.